midend/simplifyBitwise.cpp

```diff
@@ -0,0 +1,98 @@
+#include "simplifyBitwise.h"
+
+namespace P4 {
+
+bool SimplifyBitwise::Scan::preorder(const IR::AssignmentStatement *as) {
+    if (!as->right->is<IR::BOr>())
+        return false;
+    can_be_changed = true;
+    mpz_init(total_mask.get_mpz_t());
+    return true;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::BOr *bor) {
+    if (!findContext<IR::AssignmentStatement>())
+        can_be_changed = false;
+    if (!(bor->left->is<IR::BAnd>() && bor->right->is<IR::BAnd>()))
+        can_be_changed = false;
+    return can_be_changed;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::BAnd *band) {
+    if (!findContext<IR::BOr>())
+        can_be_changed = false;
+    if (!band->left->is<IR::Constant>() && !band->right->is<IR::Constant>())
+        can_be_changed = false;
+    if (band->left->is<IR::Constant>() && band->right->is<IR::Constant>())
+        can_be_changed = false;
+    return can_be_changed;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::Constant *constant) {
+    if (!findContext<IR::BAnd>()) {
+        can_be_changed = false;
+        return false;
+    }
+    // Ensure that there are no collisions within the bitmask
+    mpz_class mask;
+    mpz_init(mask.get_mpz_t());
+    mpz_class zero;
+    mpz_init(zero.get_mpz_t());
+    mpz_and(mask.get_mpz_t(), total_mask.get_mpz_t(), constant->value.get_mpz_t());
+    if (mpz_cmp(mask.get_mpz_t(), zero.get_mpz_t()) == 0) {
+        mpz_ior(total_mask.get_mpz_t(), mask.get_mpz_t(), total_mask.get_mpz_t());
+    } else {
+        can_be_changed = false;
+    }
+    return false;
+}
```

Comment:
The above mask-testing code seems wrong to me -- it will always succeed, regardless of the masks?

You should be checking that the mask are disjoint AND cover all bits of the result.  If they don't cover all bits of the result, it could still work, but you would have to assign 0 to the slices that are not covered by either mask.

You could even make it work for masks that are not disjoint by assigning to the non-disjoint slices with a BAnd.

Reply:
You could extend this with a bunch of masks if you so chose to.  I was trying to handle this corner case that was arising.  You're right, let me rework the code to get this particular type of side case.

Reply:
The important thing is to not accidentally catch other similar but not quite the same corner cases in a way that changes the semantics of the resulting code...